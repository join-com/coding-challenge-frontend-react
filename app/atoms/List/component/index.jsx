import React from 'react';
import PropTypes from 'prop-types';

const List = ({items, ...rest}) =>  (
  <ul {...rest}>
    {items.map((item, key) => <li key={key}>{item}</li>)}
  </ul>
);

List.propTypes = {
  items: PropTypes.array.isRequired
};

export default List;
