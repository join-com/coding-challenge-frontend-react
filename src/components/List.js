import React from 'react';

const List = (props) => {
	const { incidents } = props.bikes;
	if (!incidents.length) return (<p>No data found!</p>);
	return (
		<ul>
			{incidents.map((bike) => {
				let { id, title, url, description, address, occured_at, media, type, source } = bike;
				return (
					<li key={id}>
						<h3><a href={url}>{title}</a></h3>
						<img src={media.image_url_thumb} alt={description}></img>
						<dl>
							<dt>Location:</dt>
							<dd>{address}</dd>
							<dt>{type}: </dt>
							<dd>{new Date(occured_at * 1000).toLocaleDateString()}</dd>
						</dl>
						<i>Source: <a href={source.html_url}>{source.name}</a></i>
					</li>
				)
			})}
		</ul>
	)

}
export default List;