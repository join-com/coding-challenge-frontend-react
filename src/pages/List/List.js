import React, { Fragment, useState, useEffect } from 'react'
import styled from 'styled-components'
import moment from 'moment'

import CalendarRange from './CalendarRange/CalendarRange'
import SearchCaseInput from './SearchCaseInput/SearchCaseInput'
import SearchButton from './SearchButton/SearchButton'
import Card from './Card/Card'
import Pagination from './Pagination/Pagination'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`
const Content = styled.div`
  width: 100%;
`

const Filters = styled.div`
  display: flex;
  margin-top: 30px;
  margin-bottom: 30px;
`

const CountResults = styled.div`
  text-align: right;
`

const resultsPerPage = 10

const defaultState = {
  currentPageIndex: 1,
  query: '',
  dateFrom: undefined,
  dateTo: undefined,
}

function convertDateToUnixTime(date) {
  return moment(date).unix()
}

export default function List() {
  const [data, setData] = useState({ incidents: [] })
  const [countResults, setCountResults] = useState(undefined)

  const state =
    (localStorage.getItem('state') &&
      JSON.parse(localStorage.getItem('state'))) ||
    defaultState
  const [currentPageIndex, setCurrentPageIndex] = useState(
    state.currentPageIndex
  )
  const [loading, setLoading] = useState(true)
  const [triggerCount, setTriggerCount] = useState(true)
  const [query, setQuery] = useState(state.query)
  const [dateFrom, setDateFrom] = useState(state.dateFrom)
  const [dateTo, setDateTo] = useState(state.dateTo)

  useEffect(() => {
    async function fetchData() {
      if (loading) {
        const baseUrl =
          'https://bikewise.org/api/v2/incidents/?incident_type=theft&proximity=Berlin'
        let urlCount = `${baseUrl}&per_page=300` // FIXME add a mechanism to count all the results by using paging recursively
        let url = `${baseUrl}&per_page=${resultsPerPage}&page=${currentPageIndex}`

        if (dateFrom !== undefined && dateTo !== undefined) {
          url += `&occurred_after=${convertDateToUnixTime(
            dateFrom
          )}&occurred_before=${convertDateToUnixTime(dateTo)}`
          urlCount += `&occurred_after=${convertDateToUnixTime(
            dateFrom
          )}&occurred_before=${convertDateToUnixTime(dateTo)}`
        }
        if (query.length > 0) {
          url += `&query=${query}`
          urlCount += `&query=${query}`
        }
        localStorage.setItem(
          'state',
          JSON.stringify({ currentPageIndex, query, dateFrom, dateTo })
        )
        const response = await window.fetch(url)
        const results = await response.json()
        setData(results)
        setLoading(false)

        if (triggerCount) {
          const responseCount = await window.fetch(urlCount) // TODO fetch in parallels
          const resultsCount = await responseCount.json()
          setCountResults(resultsCount.incidents.length)
          setTriggerCount(false)
        }
      }
    }
    fetchData()
  }, [query, dateFrom, dateTo, currentPageIndex, loading, triggerCount])

  return (
    <Wrapper>
      <Content>
        {loading ? (
          <div>Loading...</div>
        ) : (
          <Fragment>
            <Filters>
              <SearchCaseInput
                value={query}
                onChange={event => setQuery(event.target.value)}
              />
              <CalendarRange
                from={dateFrom}
                to={dateTo}
                onDateFromChange={setDateFrom}
                onDateToChange={setDateTo}
              />
              <SearchButton
                onClick={() => {
                  setLoading(true)
                  setCurrentPageIndex(0)
                  setTriggerCount(true)
                }}
              />
            </Filters>

            {countResults === 0 ? (
              <div>No results</div>
            ) : (
              <Fragment>
                <CountResults>Total: {countResults}</CountResults>
                <div>
                  {data.incidents.map(item => (
                    <Card
                      key={item.id}
                      id={item.id}
                      title={item.title}
                      description={item.description}
                      imageUrlThumb={item.media.image_url_thumb}
                      address={item.address}
                      incidentDate={item.occurred_at}
                    />
                  ))}
                </div>
                <Pagination
                  countResults={countResults}
                  currentPageIndex={currentPageIndex}
                  resultsPerPage={resultsPerPage}
                  handleCurrentPageChange={index => {
                    setLoading(true)
                    setCurrentPageIndex(index)
                  }}
                />
              </Fragment>
            )}
          </Fragment>
        )}
      </Content>
    </Wrapper>
  )
}
