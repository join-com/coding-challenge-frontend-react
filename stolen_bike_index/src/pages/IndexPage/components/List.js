import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getAllIncidentsRequest } from "../../../redux/actions/get_all_incidents.actions";
import { countAllIncidentsRequest } from "../../../redux/actions/count_all_incidents.actions";
import { Card, Pagination, Spin } from 'antd';
import { Link } from 'react-router-dom';
import { styles } from "../styles/List.styles";
import Moment from 'react-moment';
import no_image from '../../../images/no-image-icon-15.png';
import injectSheet from 'react-jss';

const mapStateToProps = state => {
  const { incidents, loading, total, error, page, isSearch } = state.Incidents_reducer;
  return { incidents, loading, total, error, page, isSearch };
}, { Meta } = Card;

export default injectSheet(styles)(connect(mapStateToProps, null)(class List extends Component {
  componentDidMount() {
    this.props.dispatch(getAllIncidentsRequest(1));
    this.props.dispatch(countAllIncidentsRequest());
  }

  showIncidents = page => {
    !this.props.isSearch ?
      this.props.dispatch(getAllIncidentsRequest(page)) :
      this.props.dispatch({ type: 'CUSTOM_PAGINATION', page });
  };

  render() {
    const { incidents, loading, total, error, page, isSearch, classes } = this.props;
    return incidents ?
        <>
          <Pagination
            className={classes.pagination_position}
            total={total || incidents.length}
            showTotal={(total, range) => `${range[0]}-${range[1]} of ${total} incidents`}
            pageSize={10}
            onChange={this.showIncidents}
            defaultCurrent={1} />
          <Spin spinning={loading} tip="Loading...">
            {incidents.length ?
              incidents.slice(isSearch && 10 * page - 10, 10 * page).map(incident => (
                  <Card key={incident.id} hoverable className={classes.card_position}>
                    <Meta
                      avatar={incident.media.image_url_thumb ?
                        <img className={classes.img_size} src={incident.media.image_url_thumb} alt='avatar'/> :
                        <img className={classes.img_size} src={no_image} alt='no-avatar'/>}
                      title={<a href={incident.source.html_url} rel="noopener noreferrer" target='_blank'>{incident.title}</a>}
                      description={incident.description ?
                        incident.description :
                        <span className={classes.italicStyle}>No Description</span>}/>
                    <span>
                    <Moment unix format='YYYY-MM-DD HH:mm'>{incident.occurred_at}</Moment> - {incident.address}
                  </span>
                    <Link to={`/${incident.id}`} className={classes.float_right}>Details</Link>
                  </Card>
              )
            ) : <span className={classes.no_result}>No Results</span>}
          </Spin>
          </> : <span className={classes.error_message}>{error}</span>
  }
}));

