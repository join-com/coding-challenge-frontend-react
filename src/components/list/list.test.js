import React from 'react';
import { shallow , mount} from 'enzyme';
import List from './list';

describe('List', () => {
    it('it should render correctly', () => {
        const component = shallow(<List />);
        expect(component).toMatchSnapshot();
    });
})