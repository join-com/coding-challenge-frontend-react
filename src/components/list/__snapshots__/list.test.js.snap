// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List it should render correctly 1`] = `
<Connect(List)
  classes={
    Object {
      "card": "Connect-List--card-4",
      "content": "Connect-List--content-6",
      "controls": "Connect-List--controls-8",
      "cover": "Connect-List--cover-7",
      "details": "Connect-List--details-5",
      "error": "Connect-List--error-10",
      "fullPageHeight": "Connect-List--fullPageHeight-2",
      "mainContainer": "Connect-List--mainContainer-1",
      "noRecords": "Connect-List--noRecords-11",
      "playIcon": "Connect-List--playIcon-9",
      "totalCasesText": "Connect-List--totalCasesText-3",
    }
  }
/>
`;
